Bug 1200354: Add redirects test suite and move redirects to bedrock.

pmclanahan opened the pull request:
This is just the global.conf redirects so far, but it sets a good base to work from.

jgmize commented:
@pmclanahan would you mind rebasing?

pmclanahan commented:
@jgmize this is still in progress (need unit tests for a lot of this), but I wanted to get it to you now so you could digest and critique the testing methodologies and directions.

pmclanahan commented:
rebased

pmclanahan commented:
@jgmize I just added the global.conf changes. I've not yet gotten all of the redirects in global.conf, but I think enough that we can get this in as a base then move quickly to the rest.

jgmize commented on bedrock/base/urlresolvers.py:
```diff
@@ -38,6 +38,23 @@ def reverse(viewname, urlconf=None, args=None, kwargs=None, prefix=None):
+    # map with the CANONICAL_LOCALES setting.
+    langs.update((k.split('-')[0], v) for k, v in LUM.items() if
+                 k.split('-')[0] not in langs)
+    return langs
```
Does this need to be a function, or would it make more sense to construct this dict once, on module import, since it only varies with the settings are changed?

pmclanahan replied:
Made it a lazy variable so tests don't need to be modified.

jgmize commented on bedrock/base/urlresolvers.py:
```diff
@@ -93,26 +111,15 @@ def get_language(self):
-                     k.split('-')[0] not in langs)
-        try:
-            ranked = parse_accept_lang_header(accept_lang)
-        except ValueError:  # see https://code.djangoproject.com/ticket/21078
```
I'm glad we don't have to catch this exception any more. I remember thinking how it would be forever before we could remove this line, but it only took 2 years.

pmclanahan replied:
lulz. yeah. good stuff.

jgmize commented:
@pmclanahan what bug are we tracking this under?

pmclanahan commented:
Great question. I'll check that out and file if necessary.

pmclanahan commented:
Added new bug # ([bug 1200354](https://bugzilla.mozilla.org/show_bug.cgi?id=1200354)) to PR title. Will put it in the commit message as well when I rebase things.

jgmize commented:
I modified the mozorg.adhoc job on the WebQA staging server to allow running against demo instances, and found a couple of failures: http://webqa-ci-staging1.qa.scl3.mozilla.com:8080/job/mozorg.adhoc/14/console -- that server requires a VPN login to access, so I'm pasting the relevant output below:

```
00:28:33 =================================== FAILURES ===================================
00:28:33 _______________________ TestRedirects.test_dnt_redirect ________________________
00:28:33 [gw8] linux2 -- Python 2.7.9 /var/lib/jenkins/jobs/mozorg.adhoc/workspace/.env/bin/python2.7
00:28:33 self = <tests.test_redirect.TestRedirects object at 0x7f3a06e596d0>
00:28:33 mozwebqa = <pytest_mozwebqa.pytest_mozwebqa.TestSetup instance at 0x7f3a06c8c830>
00:28:33 
00:28:33     @nondestructive
00:28:33     def test_dnt_redirect(self, mozwebqa):
00:28:33         url = mozwebqa.base_url + '/dnt'
00:28:33         response = requests.get(url)
00:28:33 >       Assert.equal(response.status_code, 200)
00:28:33 
00:28:33 tests/test_redirect.py:307: 
00:28:33 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
00:28:33 
00:28:33 self = <class 'unittestzero.Assert'>, first = 404, second = 200, msg = None
00:28:33 
00:28:33     @classmethod
00:28:33     def equal(self, first, second, msg=None):
00:28:33         """
00:28:33             Asserts that 2 elements are the same
00:28:33     
00:28:33             :Args:
00:28:33              - First object to be tested
00:28:33              - Second object to be tested
00:28:33              - Message that will be printed if it fails
00:28:33             """
00:28:33 >       assert first == second, msg
00:28:33 E       AssertionError: None
00:28:33 
00:28:33 .env/lib/python2.7/site-packages/unittestzero.py:55: AssertionError
00:28:33 ---------------------------- Captured stderr setup -----------------------------
00:28:33 /var/lib/jenkins/jobs/mozorg.adhoc/workspace/.env/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py:769: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
00:28:33   InsecureRequestWarning)
00:28:33 /var/lib/jenkins/jobs/mozorg.adhoc/workspace/.env/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py:769: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
00:28:33   InsecureRequestWarning)
00:28:33 _______________________ TestRedirects.test_apps_redirect _______________________
00:28:33 [gw3] linux2 -- Python 2.7.9 /var/lib/jenkins/jobs/mozorg.adhoc/workspace/.env/bin/python2.7
00:28:33 self = <tests.test_redirect.TestRedirects object at 0x7f0da070d3d0>
00:28:33 mozwebqa = <pytest_mozwebqa.pytest_mozwebqa.TestSetup instance at 0x7f0da0718dd0>
00:28:33 
00:28:33     @nondestructive
00:28:33     def test_apps_redirect(self, mozwebqa):
00:28:33         """
00:28:33             Test mozilla.org/apps redirects to
00:28:33             http[s]://marketplace.firefox.com
00:28:33             """
00:28:33         url = mozwebqa.base_url + '/apps'
00:28:33         response = requests.get(url)
00:28:33 >       Assert.contains('marketplace.firefox.com', response.url)
00:28:33 
00:28:33 tests/test_redirect.py:253: 
00:28:33 _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
00:28:33 
00:28:33 self = <class 'unittestzero.Assert'>, needle = 'marketplace.firefox.com'
00:28:33 haystack = 'https://www-demo5.allizom.org/en-US/apps', msg = ''
00:28:33 
00:28:33     @classmethod
00:28:33     def contains(self, needle, haystack, msg=''):
00:28:33         try:
00:28:33             assert needle in haystack
00:28:33         except AssertionError:
00:28:33 >           raise AssertionError('%s is not found in %s. %s' % (needle, haystack, msg))
00:28:33 E           AssertionError: marketplace.firefox.com is not found in https://www-demo5.allizom.org/en-US/apps.
00:28:33 
00:28:33 .env/lib/python2.7/site-packages/unittestzero.py:191: AssertionError
00:28:33 ---------------------------- Captured stderr setup -----------------------------
00:28:33 /var/lib/jenkins/jobs/mozorg.adhoc/workspace/.env/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py:769: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
00:28:33   InsecureRequestWarning)
00:28:33 /var/lib/jenkins/jobs/mozorg.adhoc/workspace/.env/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py:769: InsecureRequestWarning: Unverified HTTPS request is being made. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.org/en/latest/security.html
00:28:33   InsecureRequestWarning)
00:28:33 - generated xml file: /var/lib/jenkins/jobs/mozorg.adhoc/workspace/results.xml -
00:28:33 =========================== short test summary info ============================
00:28:33 FAIL test_redirect.py::TestRedirects::()::test_dnt_redirect
00:28:33 FAIL test_redirect.py::TestRedirects::()::test_apps_redirect
00:28:33 ==================== 2 failed, 118 passed in 134.75 seconds ====================
```

jgmize commented on bedrock/mozorg/redirects.py:
```diff
@@ -0,0 +1,238 @@
+
+    # Bug 945474 - delete Marketplace marketing product page
+    # and redirect
+    redirect(r'^apps/$', 'https://marketplace.firefox.com/'),
```
The mcom-tests revealed that this only works if you include the trailing slash: `/apps/` redirects to marketplace, but `/apps` gets redirected to `/en-US/apps/` which is a 404.

jgmize commented on bedrock/mozorg/redirects.py:
```diff
@@ -0,0 +1,238 @@
+    redirect(r'^legal/eula/firefox-3/$', '/about/legal/eula/firefox-3/'),
+
+    # Bug 1073269 /dnt/ -> /firefox/dnt/
+    redirect(r'^dnt/$', 'firefox.dnt'),
```
This is the other failure that mcom-tests revealed, and has the same issue as the `/apps` redirect, and many others that we don't have coverage fore: the redirect only matches with a trailing slash.

jgmize commented:
@pmclanahan other than the trailing slash issues, this is looking pretty good. Once that's been addressed, and we get a green run from mcom-tests against a demo deploy, and you squash the commits, I think we'll be good to merge. r+wc, nice work.

pmclanahan commented:
@jgmize assuming the tests pass, are we ready for merge? I just rebased and squashed a bit. Still kindof a large number of commits, but I think they're all kindof okay. Would you prefer more squarshing?

jgmize commented on bedrock/mozorg/redirects.py:
```diff
@@ -0,0 +1,238 @@
+             'https://developer.mozilla.org/docs/Python_binding_for_NSS'),
+
+    # bug 1043035
+    redirect('^projects/security/pki/(?:index\.html)?$',
```
This looks like it still requires a trailing slash.
